crates/bevy_ecs/src/query/iter.rs

```diff
@@ -1151,35 +1152,42 @@ where
             tables: &world.storages().tables,
             fetch,
             entity_iter: entity_list.into_iter(),
+            previous_archetype_id: ArchetypeId::INVALID,
         }
     }
 
     /// # Safety
     /// `entity` must stem from `self.entity_iter`, and not have been passed before.
     #[inline(always)]
     unsafe fn fetch_next(&mut self, entity: Entity) -> D::Item<'w> {
-        let (location, archetype, table);
-        // SAFETY:
-        // `tables` and `archetypes` belong to the same world that the [`QueryIter`]
-        // was initialized for.
-        unsafe {
-            location = self.entities.get(entity).debug_checked_unwrap();
-            archetype = self
-                .archetypes
-                .get(location.archetype_id)
-                .debug_checked_unwrap();
-            table = self.tables.get(location.table_id).debug_checked_unwrap();
-        }
+        let location = self.entities.get(entity).debug_checked_unwrap();
 
-        // SAFETY: `archetype` is from the world that `fetch` was created for,
-        // `fetch_state` is the state that `fetch` was initialized with
-        unsafe {
-            D::set_archetype(
-                &mut self.fetch,
-                &self.query_state.fetch_state,
-                archetype,
-                table,
-            );
+        if self.previous_archetype_id != location.archetype_id {
+            self.previous_archetype_id = location.archetype_id;
```

Comment:
Do we need to worry about unwind safety?  If `set_archetype` panics and the caller does a `catch_unwind` and then calls this again, we'll call `fetch` without `set_archetype` having completed.  

If we move the assignment after the call to `set_archetype`, then there's the risk that we'll call `fetch` for an entity in the *previous* archetype without a fresh `set_archetype` call, which could also be bad.  

So maybe we have to set `previous_archetype_id` to `INVALID` here and then to `location.archetype_id` after the call to `set_archetype`?  

Or is something in here already completely unwind-unsafe and we don't need to worry about making it worse?

Reply:
Do we? Going by unwind safety, `QueryIterationCursor` can skip over storage ids, which would make several iterator types behave incorrectly.

Reply:
Yeah, but I don't think skipping entities can cause any unsoundness, since we don't impl `TrustedLen` anywhere.  

We definitely have `QueryData` where it's unsound to call `fetch` without calling `set_table` or `set_archetype` first.  So if you have some `EvilQueryData` that panics on one table and then works on the next, then a `Query<(EvilQueryData, &C)>` could cause UB when `<&C>::fetch` indexes into the wrong table column.  

We might be able to blame the `unsafe impl WorldQuery for EvilQueryData`, but `WorldQuery` doesn't currently require that `set_table` and `set_archetype` not panic.  

I'd be gently in favor of setting it to `INVALID` during the calls just to be safe, but I admit it's a crazy edge case that's unlikely to come up, so I'm not going to push too hard.  :)

Reply:
hmm, `Query::iter` explicitly states:
"This iterator is always guaranteed to return results from each matching entity once and only once." (No # Panic section)
Breaking this could cause UB in code that relies on this.
Besides, storage id skipping could allow `EvilQueryData` to falsify query emptiness, `single` and `Populated` checks, which are also present in `get_param`/`validate_param`. It could break wide swaths of bevy machinery!
Maybe we want to adjust the Safety section on `WorldQuery`?
I'd be in favor of making a note comment for now, and if we want unwind safety, adding it later.

Reply:
under what conditions does set_archetype panic?

Reply:
It is a trait function (with no Safety requirement regarding panicking), so there is no guarantee that it won't when implemented by some library/user. For in-engine impls `set_archetype` shouldn't panic.